stat/roc.go

```diff
@@ -126,3 +126,32 @@ func ROC(cutoffs, y []float64, classes []bool, weights []float64) (tpr, fpr, thr
 
 	return tpr, fpr, cutoffs
 }
+
+// TOC returns the Total Operating Characteristic for the classes provided,
+// which should be a boolean variable ordered by a rank variable. TOC also
+// returns the minimum and maximum bounds for the TOC.
+//
+// More details about TOC curves are available at
+// https://en.wikipedia.org/wiki/Total_operating_characteristic
+func TOC(classes []bool) (min, toc, max []float64) {
+	if len(classes) == 0 {
+		return nil, nil, nil
+	}
+	toc = make([]float64, len(classes)+1)
+	for i := range toc[1:] {
+		toc[i+1] = toc[i]
+		if classes[i] {
+			toc[i+1] += 1
+		}
+	}
+	last := toc[len(toc)-1]
+	min = make([]float64, len(toc))
+	for i := range min {
+		min[i] = math.Max(0, last+float64(i-len(classes)))
```

Comment:
why do this as a float operation rather than an int operation? These should both be int's. That said, if we add functionality to allow weights, then this will need to be a float operation but in that case it would be a different total not `len(classes)'.

Reply:
Also, here `last + i - len(classes)' would make more sense to me as `i - (len(classes) - last)' I realise this is doing slightly more work (?) but just though I'd mention it has a more obvious meaning. Maybe a comment here explaining that `last' is actually the number of positives in the data, and so `len(classes) - last' is the number of negatives in the data, while i is the number of data assigned positive.

Reply:
It's floats because it will get used as floats later and we don't have (thankfully) vector coercion.

Reply:
I've reorganised, but repeated reconversion is not a thing I want to do.

Reply:
that's fair, especially since we'll probably introduce weights. Without weights, all three outputs (min, toc, max) could all be []int though at least as far as the scope of this function is concerned, but that's probably not important.